src/model_repository_manager.cc

```diff
@@ -1623,4 +1607,284 @@ ModelRepositoryManager::CheckNode(
+  updated_node->upstreams_.clear();
+  updated_node->missing_upstreams_.clear();
+  updated_node->connected_ = true;
+  if (updated_node->model_config_.has_ensemble_scheduling()) {
+    auto mutable_steps = updated_node->model_config_.mutable_ensemble_scheduling()->mutable_step();
+    for (auto& step : *mutable_steps) {
+      // Build model identifier to look for the composing model,
+      // currently using the same namespace as the 'updated_node' for
+      // preferred identifier.
+      // [enhancement] May allow user-specified namespace once allowed in
+      // ensemble config.
+      const ModelIdentifier preferred_id(updated_node->model_id_.namespace_, step.model_name());
+      auto node = FindNode(preferred_id, true /* allow_fuzzy_matching */);
+      if (node) {
+        // Set 'model_namespace' field in model config to allow direct model
+        // lookup at later stage.
+        step.set_model_namespace(node->model_id_.namespace_);
+
+        // Add the current node to its upstream
+        node->downstreams_.emplace(updated_node);
+
+        // Add the upstream to the current node
+        auto res = updated_node->upstreams_.emplace(
+          node, std::set<int64_t>({step.model_version()}));
+        // If map insertion doesn't happen, the same model is required in
+        // different step, insert the version to existing required version set.
+        if (!res.second) {
+          res.first->second.insert(step.model_version());
+        }
+      } else {
+        updated_node->connected_ = false;
+        updated_node->missing_upstreams_.emplace(step.model_name());
+      }
+
+      // If no node is found or the found node is not exact matched,
+      // just record that in 'missing_nodes_' so the current node will
+      // be rechecked whenever a new candidate is added.
+      if (!node || (node->model_id_ != preferred_id)) {
+        missing_nodes_[preferred_id.name_].emplace(updated_node->model_id_);
+      }
+
+      // Only update status if there is no failure from earlier stages
+      if (updated_node->status_.IsOk() && !updated_node->connected_) {
+        std::string name_list;
+        for (const auto& missing_name : updated_node->missing_upstreams_) {
+          if (!name_list.empty()) {
+            name_list += ", ";
+          }
+          name_list += missing_name;
+        }
+        updated_node->status_ = Status(
+            Status::Code::INVALID_ARG,
+            "ensemble " + updated_node->model_id_.str() +
+                " contains models that are not available or ambiguous: " + name_list);
+      }
+    }
+  }
+}
+
+std::pair<std::set<ModelIdentifier>, std::set<ModelIdentifier>>
+ModelRepositoryManager::DependencyGraph::RemoveNode(const ModelIdentifier& model_id)
+{
+  std::set<ModelIdentifier> upstreams, downstreams;
+  auto it = nodes_.find(model_id);
+  // no-op if not found: "node" has been removed
+  if (it != nodes_.end()) {
```

Comment:
May be able to simplify code by early-exiting if not found. To avoid indenting if block below. Totally up to you, feel free to hit resolve if you'd rather not do this.